src/main/scala/higherkindness/rules_scala/common/worker/FutureTaskWaitOnCancel.scala

```diff
@@ -0,0 +1,69 @@
+package higherkindness.rules_scala
+package common.worker
+
+import java.util.concurrent.{Callable, CancellationException, FutureTask, TimeUnit}
+import java.util.concurrent.locks.ReentrantLock
+
+/**
+ * This is a FutureTask that, when cancelled, waits for its callable to end, either by interruption or by completing.
```

Comment:
Could we replicate this sort of thing by combining a `Promise` that constitutes a cancellation signal with the `Future` that does the work? E.g. something like:
```scala
class CancellableFuture[+A](underlying: Future[+A]) {
  val cancel: Promise[Unit] = Promise()
  val result = underlying.zip(cancel).map { case (result, _) => result }(ExecutionContext.parasitic)
}
```

Reply:
I don't think this works because Future itself has no mechanism for cancelling, so the underlying callable or thread is never interrupted. Meaning you can mark it as interrupted, but the thread keeps on chugging away. 

I'm also not sure how I would get access to the underlying callable/runnable here to wait on it completing.

The cancel method on FutureTask provides me the ability to mark the ongoing thread as interrupted. Then using wait for that to complete using the custom Callable.